chat-android/src/main/java/com/ably/chat/Typing.kt

```diff
@@ -119,130 +129,278 @@ internal class DefaultTyping(
+        } catch (e: Exception) {
+            logger.error("DefaultTyping.sendTyping(); failed to publish typing event", e)
+            throw e
+        }
     }
 
     override fun release() {
-        presenceSubscription.unsubscribe()
+        typingEventPubSubSubscription.unsubscribe()
         typingScope.cancel()
         room.realtimeClient.channels.release(channelWrapper.name)
     }
 
-    private fun startTypingTimer() {
-        val timeout = room.options.typing?.timeoutMs ?: throw AblyException.fromErrorInfo(
-            ErrorInfo(
-                "Typing options hasn't been initialized",
-                ErrorCode.BadRequest.code,
-            ),
-        )
+    private fun startHeartbeatTimer() {
         logger.trace("DefaultTyping.startTypingTimer()")
-        typingJob = typingScope.launch {
-            delay(timeout)
-            logger.debug("DefaultTyping.startTypingTimer(); timeout expired")
-            stop()
+        typingHeartBeatTimerJob = typingScope.launch {
+            delay(heartbeatThrottle)
+            logger.debug("DefaultTyping.startTypingTimer(); heartbeatThrottle expired")
         }
     }
 
-    private suspend fun processEvent() {
-        var numRetries = 0
-        while (numRetries <= PRESENCE_GET_MAX_RETRIES) {
-            try {
-                val currentlyTyping = get()
-                emit(currentlyTyping)
-                return // Exit if successful
-            } catch (e: Exception) {
-                numRetries++
-                val delayDuration = min(
-                    PRESENCE_GET_RETRY_MAX_INTERVAL_MS,
-                    PRESENCE_GET_RETRY_INTERVAL_MS * 2.0.pow(numRetries).toLong(),
-                )
-                logger.debug("Retrying in $delayDuration ms... (Attempt $numRetries of $PRESENCE_GET_MAX_RETRIES)", e)
-                delay(delayDuration)
+    /**
+     * Adds a typing event to [typingEvents].
+     *
+     * It will cancel the timed death of an typing event with the same key (userId)
+     * if such exists.
+     *
+     * @param eventType The typing event will be added to [typingEvents].
+     * @param clientId The ID of the user tied to the typing event.
+     * Spec: CHA-T13b
+     */
+    private fun processEvent(eventType: TypingEventType, clientId: String) {
+        when (eventType) {
+            TypingEventType.Started -> { // CHA-T13b1
+                currentlyTypingMembers.add(clientId)
+                // CHA-T13b2 - Cancel the current self stopping event and replace with new one
+                typingEvents[clientId]?.cancelJob()
+                // CHA-T10a1 - If a typing.start not received within this period, the client shall assume that the user has stopped typing
+                val typingEventWaitingTimeout = heartbeatThrottle + timeoutMs
+                // CHA-T13b3
+                val timedTypingStartEvent = TimedTypingStartEventPruner(typingScope, clientId, typingEventWaitingTimeout) {
```

Comment:
I am not sure that 

```kotlin
TimedTypingStartEventPruner(typingScope, clientId, typingEventWaitingTimeout) {
         // typingEventWaitingTimeout elapsed, so remove given clientId and emit stop event
         currentlyTypingMembers.remove(it)
         typingEvents.remove(it)
         emit(TypingEventType.Stopped, it)
}
```

is more readable than:

```kotlin
typingScope.launch {
         // typingEventWaitingTimeout elapsed, so remove given clientId and emit stop event
         delay(waitingPeriod)
         removeTypingEvent(clientId)
         currentlyTypingMembers.remove(it)
         typingEvents.remove(it)
         emit(TypingEventType.Stopped, it)
}
```
And new abstraction is necessary

Reply:
Fixed https://github.com/ably/ably-chat-kotlin/pull/117/commits/0886eec9584e082a8bc5f221f344557ecc6ec078

Reply:
Earlier the class had good amount of code, later on we removed lot of stuff, as of now, we don't need it, it seems